File: ios/EMWDATModule.swift
import ExpoModulesCore
import MWDATCore
import MWDATCamera
import MWDATDisplay

public class EMWDATModule: Module {
    private let logger = EMWDATLogger.shared

    public func definition() -> ModuleDefinition {
        Name("EMWDAT")

        Events(
            "onRegistrationStateChange",
            "onDevicesChange",
            "onLinkStateChange",
            "onDeviceStateChange",
            "onStreamStateChange",
            "onCameraStateChange",
            "onVideoFrame",
            "onPhotoCaptured",
            "onStreamError",
            "onPermissionStatusChange",
            "onCompatibilityChange",
            "onDeviceSessionStateChange",
            "onDeviceSessionError",
            "onCapabilityStateChange",
            "onDisplayStateChange",
            "onDisplayTap",
            "onDisplayError",
            "onDisplayVideoEvent"
        )

        // MARK: - Lifecycle

        OnCreate {
            self.logger.info("Module", "Module created")
            Task { @MainActor in
                let emitter: EventEmitter = { [weak self] name, body in
                    self?.sendEvent(name, body)
                }
                WearablesManager.shared.setEventEmitter(emitter)
                CameraSessionManager.shared.setEventEmitter(emitter)
                DisplayManager.shared.setEventEmitter(emitter)
            }
        }

        OnDestroy {
            self.logger.info("Module", "Module destroyed")
            Task { @MainActor in
                CameraSessionManager.shared.destroy()
                DisplayManager.shared.destroy()
                WearablesManager.shared.cleanup()
            }
        }

        // MARK: - Logging

        Function("setLogLevel") { (level: String) in
            let logLevel: EMWDATLogLevel
            switch level {
            case "debug": logLevel = .debug
            case "info": logLevel = .info
            case "warn": logLevel = .warn
            case "error": logLevel = .error
            case "none": logLevel = .none
            default: logLevel = .info
            }
            self.logger.setLogLevel(logLevel)
            self.logger.info("Module", "Log level set", context: ["level": level])
        }

        // MARK: - Configuration

        AsyncFunction("configure") { (promise: Promise) in
            Task { @MainActor in
                do {
                    try WearablesManager.shared.configure()
                    promise.resolve(nil)
                } catch let error as WearablesError {
                    promise.reject("CONFIGURATION_FAILED", self.describeWearablesError(error))
                } catch {
                    promise.reject("CONFIGURATION_FAILED", "Failed to configure SDK: \(error.localizedDescription)")
                }
            }
        }

        // MARK: - Registration

        Function("getRegistrationState") { () -> String in
            return "unavailable"
        }

        AsyncFunction("getRegistrationStateAsync") { (promise: Promise) in
            Task { @MainActor in
                let state = WearablesManager.shared.currentRegistrationState
                promise.resolve(self.mapRegistrationState(state))
            }
        }

        AsyncFunction("startRegistration") { (promise: Promise) in
            Task { @MainActor in
                do {
                    try await WearablesManager.shared.startRegistration()
                    promise.resolve(nil)
                } catch let error as RegistrationError {
                    promise.reject("REGISTRATION_FAILED", self.describeRegistrationError(error))
                } catch {
                    promise.reject("REGISTRATION_FAILED", error.localizedDescription)
                }
            }
        }

        AsyncFunction("startUnregistration") { (promise: Promise) in
            Task { @MainActor in
                do {
                    try await WearablesManager.shared.startUnregistration()
                    promise.resolve(nil)
                } catch let error as UnregistrationError {
                    promise.reject("UNREGISTRATION_FAILED", self.describeUnregistrationError(error))
                } catch {
                    promise.reject("UNREGISTRATION_FAILED", error.localizedDescription)
                }
            }
        }

        // MARK: - URL Handling

        AsyncFunction("handleUrl") { (url: String, promise: Promise) in
            guard let parsedUrl = URL(string: url) else {
                self.logger.warn("Module", "Invalid URL", context: ["url": url])
                promise.resolve(false)
                return
            }
            Task { @MainActor in
                do {
                    let handled = try await Wearables.shared.handleUrl(parsedUrl)
                    promise.resolve(handled)
                } catch let error as WearablesHandleURLError {
                    switch error {
                    case .registrationError:
                        promise.reject("HANDLE_URL_REGISTRATION_ERROR", "\(error)")
                    case .unregistrationError:
                        promise.reject("HANDLE_URL_UNREGISTRATION_ERROR", "\(error)")
                    @unknown default:
                        promise.reject("HANDLE_URL_ERROR", "\(error)")
                    }
                } catch {
                    self.logger.error("Module", "handleUrl failed", error: error)
                    promise.resolve(false)
                }
            }
        }

        // MARK: - Permissions

        AsyncFunction("checkPermissionStatus") { (permission: String, promise: Promise) in
            guard permission == "camera" else {
                promise.resolve("denied")
                return
            }
            Task { @MainActor in
                do {
                    let status = try await WearablesManager.shared.checkPermissionStatus(.camera)
                    promise.resolve(self.mapPermissionStatus(status))
                } catch {
                    self.logger.error("Module", "checkPermissionStatus failed", error: error)
                    promise.resolve("denied")
                }
            }
        }

        AsyncFunction("requestPermission") { (permission: String, promise: Promise) in
            guard permission == "camera" else {
                promise.reject("INVALID_PERMISSION", "Unknown permission: \(permission)")
                return
            }
            Task { @MainActor in
                do {
                    let status = try await WearablesManager.shared.requestPermission(.camera)
                    promise.resolve(self.mapPermissionStatus(status))
                } catch {
                    promise.reject("PERMISSION_FAILED", error.localizedDescription)
                }
            }
        }

        // MARK: - Devices

        AsyncFunction("getDevices") { (promise: Promise) in
            Task { @MainActor in
                promise.resolve(WearablesManager.shared.getDevices())
            }
        }

        AsyncFunction("getDevice") { (identifier: String, promise: Promise) in
            Task { @MainActor in
                promise.resolve(WearablesManager.shared.getDevice(identifier: identifier))
            }
        }

        AsyncFunction("openFirmwareUpdate") { (promise: Promise) in
            Task { @MainActor in
                do {
                    try await WearablesManager.shared.openFirmwareUpdate()
                    promise.resolve(nil)
                } catch let error as NavigationError {
                    promise.reject("NAVIGATION_FAILED", error.description)
                } catch {
                    promise.reject("NAVIGATION_FAILED", error.localizedDescription)
                }
            }
        }

        AsyncFunction("openDATGlassesAppUpdate") { (promise: Promise) in
            Task { @MainActor in
                do {
                    try await WearablesManager.shared.openDATGlassesAppUpdate()
                    promise.resolve(nil)
                } catch let error as NavigationError {
                    promise.reject("NAVIGATION_FAILED", error.description)
                } catch {
                    promise.reject("NAVIGATION_FAILED", error.localizedDescription)
                }
            }
        }

        // MARK: - Session Management

        AsyncFunction("createSession") { (deviceId: String?, promise: Promise) in
            Task { @MainActor in
                do {
                    let sessionId = try WearablesManager.shared.createSession(deviceId: deviceId)
                    promise.resolve(sessionId)
                } catch {
                    promise.reject("SESSION_CREATE_FAILED", error.localizedDescription)
                }
            }
        }

        AsyncFunction("startSession") { (sessionId: String, promise: Promise) in
            Task { @MainActor in
                do {
                    try WearablesManager.shared.startSession(sessionId: sessionId)
                    promise.resolve(nil)
                } catch {
                    promise.reject("SESSION_START_FAILED", error.localizedDescription)
                }
            }
        }

        AsyncFunction("stopSession") { (sessionId: String, promise: Promise) in
            Task { @MainActor in
                do {
                    try WearablesManager.shared.stopSession(sessionId: sessionId)
                    promise.resolve(nil)
                } catch {
                    promise.reject("SESSION_STOP_FAILED", error.localizedDescription)
                }
            }
        }

        AsyncFunction("addCameraToSession") { (sessionId: String, config: [String: Any], promise: Promise) in
            Task { @MainActor in
                do {
                    let streamConfig = CameraSessionManager.parseConfig(from: config)
                    try CameraSessionManager.shared.addCameraToSession(sessionId: sessionId, config: streamConfig)
                    promise.resolve(nil)
                } catch let error as DeviceSessionError {
                    promise.reject("CAMERA_ADD_FAILED", error.description)
                } catch {
                    promise.reject("CAMERA_ADD_FAILED", error.localizedDescription)
                }
            }
        }

        AsyncFunction("removeCameraFromSession") { (sessionId: String, promise: Promise) in
            Task { @MainActor in
                CameraSessionManager.shared.removeCameraFromSession(sessionId: sessionId)
                promise.resolve(nil)
            }
        }

        // MARK: - Display

        AsyncFunction("addDisplayToSession") { (sessionId: String, promise: Promise) in
            Task { @MainActor in
                do {
                    try DisplayManager.shared.addDisplayToSession(sessionId: sessionId)
                    promise.resolve(nil)
                } catch let error as DeviceSessionError {
                    promise.reject("DISPLAY_ADD_FAILED", error.description)
                } catch {
                    promise.reject("DISPLAY_ADD_FAILED", error.localizedDescription)
                }
            }
        }

        AsyncFunction("renderDisplay") { (sessionId: String, root: [String: Any], promise: Promise) in
            Task { @MainActor in
                do {
                    try await DisplayManager.shared.renderDisplay(sessionId: sessionId, root: root)
                    promise.resolve(nil)
                } catch let error as DisplayError {
                    DisplayManager.shared.emitDisplayError(sessionId: sessionId, error: error)
                    promise.reject("DISPLAY_RENDER_FAILED", error.description)
                } catch {
                    promise.reject("DISPLAY_RENDER_FAILED", error.localizedDescription)
                }
            }
        }

        AsyncFunction("clearDisplay") { (sessionId: String, promise: Promise) in
            Task { @MainActor in
                do {
                    try await DisplayManager.shared.clearDisplay(sessionId: sessionId)
                    promise.resolve(nil)
                } catch let error as DisplayError {
                    DisplayManager.shared.emitDisplayError(sessionId: sessionId, error: error)
                    promise.reject("DISPLAY_CLEAR_FAILED", error.description)
                } catch {
                    promise.reject("DISPLAY_CLEAR_FAILED", error.localizedDescription)
                }
            }
        }

        AsyncFunction("removeDisplayFromSession") { (sessionId: String, promise: Promise) in
            Task { @MainActor in
                DisplayManager.shared.removeDisplayFromSession(sessionId: sessionId)
                promise.resolve(nil)
            }
        }

        AsyncFunction("getDisplayState") { (sessionId: String, promise: Promise) in
            Task { @MainActor in
                promise.resolve(DisplayManager.shared.getDisplayState(sessionId: sessionId) ?? "stopped")
            }
        }

        // MARK: - Photo Capture

        AsyncFunction("capturePhoto") { (format: String, promise: Promise) in
            Task { @MainActor in
                let photoFormat: PhotoCaptureFormat = format == "heic" ? .heic : .jpeg
                let success = CameraSessionManager.shared.capturePhoto(format: photoFormat)
                if success {
                    promise.resolve(nil)
                } else {
                    promise.reject("CAPTURE_FAILED", "Failed to capture photo - stream may not be active")
                }
            }
        }

        // MARK: - Mock Device Kit (DEBUG only)

        #if DEBUG
        AsyncFunction("enableMockDeviceKit") { (config: [String: Any], promise: Promise) in
            Task { @MainActor in
                let initiallyRegistered = config["initiallyRegistered"] as? Bool ?? true
                let initialPermissionsGranted = config["initialPermissionsGranted"] as? Bool ?? true
                MockDeviceManager.shared.enableMockDeviceKit(
                    initiallyRegistered: initiallyRegistered,
                    initialPermissionsGranted: initialPermissionsGranted
                )
                promise.resolve(nil)
            }
        }

        AsyncFunction("disableMockDeviceKit") { (promise: Promise) in
            Task { @MainActor in
                MockDeviceManager.shared.disableMockDeviceKit()
                promise.resolve(nil)
            }
        }

        AsyncFunction("isMockDeviceKitEnabled") { (promise: Promise) in
            Task { @MainActor in
                promise.resolve(MockDeviceManager.shared.isMockDeviceKitEnabled())
            }
        }

        AsyncFunction("pairMockDevice") { (model: String?, promise: Promise) in
            Task { @MainActor in
                do {
                    let id = try MockDeviceManager.shared.pairMockDevice(model: model ?? "rayBanMeta")
                    promise.resolve(id)
                } catch {
                    promise.reject("MOCK_DEVICE_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("unpairMockDevice") { (deviceId: String, promise: Promise) in
            Task { @MainActor in
                do {
                    try MockDeviceManager.shared.unpairMockDevice(id: deviceId)
                    promise.resolve(nil)
                } catch {
                    promise.reject("MOCK_DEVICE_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("getMockDevices") { (promise: Promise) in
            Task { @MainActor in
                promise.resolve(MockDeviceManager.shared.getMockDevices())
            }
        }

        AsyncFunction("mockDevicePowerOn") { (id: String, promise: Promise) in
            Task { @MainActor in
                do {
                    try MockDeviceManager.shared.powerOn(id: id)
                    promise.resolve(nil)
                } catch {
                    promise.reject("MOCK_DEVICE_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("mockDevicePowerOff") { (id: String, promise: Promise) in
            Task { @MainActor in
                do {
                    try MockDeviceManager.shared.powerOff(id: id)
                    promise.resolve(nil)
                } catch {
                    promise.reject("MOCK_DEVICE_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("mockDeviceDon") { (id: String, promise: Promise) in
            Task { @MainActor in
                do {
                    try MockDeviceManager.shared.don(id: id)
                    promise.resolve(nil)
                } catch {
                    promise.reject("MOCK_DEVICE_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("mockDeviceDoff") { (id: String, promise: Promise) in
            Task { @MainActor in
                do {
                    try MockDeviceManager.shared.doff(id: id)
                    promise.resolve(nil)
                } catch {
                    promise.reject("MOCK_DEVICE_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("mockDeviceFold") { (id: String, promise: Promise) in
            Task { @MainActor in
                do {
                    try MockDeviceManager.shared.fold(id: id)
                    promise.resolve(nil)
                } catch {
                    promise.reject("MOCK_DEVICE_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("mockDeviceUnfold") { (id: String, promise: Promise) in
            Task { @MainActor in
                do {
                    try MockDeviceManager.shared.unfold(id: id)
                    promise.resolve(nil)
                } catch {
                    promise.reject("MOCK_DEVICE_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("mockDeviceTap") { (id: String, promise: Promise) in
            Task { @MainActor in
                do {
                    try MockDeviceManager.shared.tap(id: id)
                    promise.resolve(nil)
                } catch {
                    promise.reject("MOCK_DEVICE_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("mockDeviceTapAndHold") { (id: String, promise: Promise) in
            Task { @MainActor in
                do {
                    try MockDeviceManager.shared.tapAndHold(id: id)
                    promise.resolve(nil)
                } catch {
                    promise.reject("MOCK_DEVICE_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("mockDeviceSetCameraFeed") { (id: String, fileUrl: String, promise: Promise) in
            let url: URL
            if fileUrl.hasPrefix("file://") {
                let path = String(fileUrl.dropFirst("file://".count))
                url = URL(fileURLWithPath: path)
            } else if let parsed = URL(string: fileUrl) {
                url = parsed
            } else {
                promise.reject("MOCK_DEVICE_ERROR", "Invalid file URL: \(fileUrl)")
                return
            }
            Task { @MainActor in
                do {
                    try MockDeviceManager.shared.setCameraFeed(id: id, fileURL: url)
                    promise.resolve(nil)
                } catch {
                    promise.reject("MOCK_DEVICE_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("mockDeviceSetCapturedImage") { (id: String, fileUrl: String, promise: Promise) in
            let url: URL
            if fileUrl.hasPrefix("file://") {
                let path = String(fileUrl.dropFirst("file://".count))
                url = URL(fileURLWithPath: path)
            } else if let parsed = URL(string: fileUrl) {
                url = parsed
            } else {
                promise.reject("MOCK_DEVICE_ERROR", "Invalid file URL: \(fileUrl)")
                return
            }
            Task { @MainActor in
                do {
                    try MockDeviceManager.shared.setCapturedImage(id: id, fileURL: url)
                    promise.resolve(nil)
                } catch {
                    promise.reject("MOCK_DEVICE_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("mockDeviceSetCameraFeedFromCamera") { (id: String, facing: String, promise: Promise) in
            Task { @MainActor in
                do {
                    try MockDeviceManager.shared.setCameraFeedFromCamera(id: id, facing: facing)
                    promise.resolve(nil)
                } catch {
                    promise.reject("MOCK_DEVICE_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("mockSetPermissionStatus") { (permission: String, status: String, promise: Promise) in
            Task { @MainActor in
                MockDeviceManager.shared.setPermissionStatus(permission: permission, status: status)
                promise.resolve(nil)
            }
        }

        AsyncFunction("mockSetPermissionRequestResult") { (permission: String, result: String, promise: Promise) in
            Task { @MainActor in
                MockDeviceManager.shared.setPermissionRequestResult(permission: permission, result: result)
                promise.resolve(nil)
            }
        }
        #endif

        // MARK: - View

        View(EMWDATStreamView.self) {
            Prop("isActive") { (view: EMWDATStreamView, isActive: Bool) in
                view.setActive(isActive)
            }

            Prop("resizeMode") { (view: EMWDATStreamView, resizeMode: String) in
                view.setResizeMode(resizeMode)
            }
        }
    }

    // MARK: - Mapping Helpers

    private func mapRegistrationState(_ state: RegistrationState) -> String {
        switch state {
        case .unavailable: return "unavailable"
        case .available: return "available"
        case .registering: return "registering"
        case .registered: return "registered"
        @unknown default: return "unavailable"
        }
    }

    private func mapPermissionStatus(_ status: PermissionStatus) -> String {
        switch status {
        case .granted: return "granted"
        case .denied: return "denied"
        @unknown default: return "denied"
        }
    }

    // MARK: - Error Descriptions

    private func describeWearablesError(_ error: WearablesError) -> String {
        switch error {
        case .internalError:
            return "Internal SDK error during configuration."
        case .alreadyConfigured:
            return "Wearables SDK is already configured."
        case .configurationError:
            return "SDK configuration error. Check Info.plist MWDAT dictionary (MetaAppID, ClientToken, AppLinkURLScheme)."
        @unknown default:
            return "Unexpected SDK error."
        }
    }

    private func describeRegistrationError(_ error: RegistrationError) -> String {
        switch error {
        case .alreadyRegistered:
            return "Device is already registered."
        case .configurationInvalid:
            return "MWDAT configuration is invalid. Check Info.plist MWDAT dictionary (MetaAppID, ClientToken, AppLinkURLScheme). Ensure Developer Mode is enabled in the Meta AI app."
        case .metaAINotInstalled:
            return "Meta AI app is not installed on this device."
        case .networkUnavailable:
            return "Network is unavailable."
        case .timeout:
            return "Registration timed out."
        case .unknown:
            return "Unknown registration error."
        @unknown default:
            return "Unexpected registration error."
        }
    }

    private func describeUnregistrationError(_ error: UnregistrationError) -> String {
        switch error {
        case .alreadyUnregistered:
            return "Device is already unregistered."
        case .configurationInvalid:
            return "MWDAT configuration is invalid. Check Info.plist MWDAT dictionary."
        case .metaAINotInstalled:
            return "Meta AI app is not installed on this device."
        case .timeout:
            return "Unregistration timed out."
        case .unknown:
            return "Unknown unregistration error."
        @unknown default:
            return "Unexpected unregistration error."
        }
    }
}

File: android/src/main/java/expo/modules/emwdat/EMWDATModule.kt
package expo.modules.emwdat

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

class EMWDATModule : Module() {
    private val logger = EMWDATLogger
    private val moduleScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private val isDebug by lazy {
        try {
            val appContext = appContext.reactContext?.applicationContext ?: return@lazy false
            val appInfo = appContext.applicationInfo
            appInfo.flags and android.content.pm.ApplicationInfo.FLAG_DEBUGGABLE != 0
        } catch (e: Exception) {
            false
        }
    }

    override fun definition() = ModuleDefinition {
        Name("EMWDAT")

        Events(
            "onRegistrationStateChange",
            "onDevicesChange",
            "onLinkStateChange",
            "onDeviceStateChange",
            "onStreamStateChange",
            "onCameraStateChange",
            "onVideoFrame",
            "onPhotoCaptured",
            "onStreamError",
            "onPermissionStatusChange",
            "onCompatibilityChange",
            "onDeviceSessionStateChange",
            "onDeviceSessionError",
            "onCapabilityStateChange",
            "onDisplayStateChange",
            "onDisplayTap",
            "onDisplayError",
            "onDisplayVideoEvent"
        )

        OnCreate {
            logger.info("Module", "Module created")
            val emitter: EventEmitter = { name, body ->
                sendEvent(name, body)
            }
            WearablesManager.setEventEmitter(emitter)
            WearablesManager.setScope(moduleScope)
            CameraSessionManager.setEventEmitter(emitter)
            CameraSessionManager.setScope(moduleScope)
            DisplayManager.setEventEmitter(emitter)
            DisplayManager.setScope(moduleScope)
        }

        OnDestroy {
            logger.info("Module", "Module destroyed")
            CameraSessionManager.destroy()
            DisplayManager.destroy()
            WearablesManager.cleanup()
            moduleScope.cancel()
        }

        // MARK: - Logging

        Function("setLogLevel") { level: String ->
            val logLevel = EMWDATLogLevel.fromString(level)
            EMWDATLogger.setLogLevel(logLevel)
            logger.info("Module", "Log level set", mapOf("level" to level))
        }

        // MARK: - Configuration

        AsyncFunction("configure") {
            val context = appContext.reactContext?.applicationContext
                ?: throw Exception("Application context not available")

            // Request BLUETOOTH_CONNECT runtime permission on Android 12+
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                val activity = appContext.currentActivity
                if (activity != null && ContextCompat.checkSelfPermission(
                        activity, Manifest.permission.BLUETOOTH_CONNECT
                    ) != PackageManager.PERMISSION_GRANTED
                ) {
                    ActivityCompat.requestPermissions(
                        activity,
                        arrayOf(Manifest.permission.BLUETOOTH_CONNECT),
                        1001
                    )
                }
            }

            WearablesManager.configure(context)
        }

        // MARK: - Registration

        Function("getRegistrationState") {
            WearablesManager.currentRegistrationState
        }

        AsyncFunction("getRegistrationStateAsync") {
            WearablesManager.currentRegistrationState
        }

        AsyncFunction("startRegistration") {
            val activity = appContext.currentActivity
                ?: throw Exception("Current activity not available")
            WearablesManager.startRegistration(activity)
        }

        AsyncFunction("startUnregistration") {
            val activity = appContext.currentActivity
                ?: throw Exception("Current activity not available")
            WearablesManager.startUnregistration(activity)
        }

        // MARK: - URL Handling (no-op on Android — deep links handled via intent-filter)

        AsyncFunction("handleUrl") { _: String ->
            false
        }

        // MARK: - Permissions

        AsyncFunction("checkPermissionStatus") { permission: String ->
            if (permission != "camera") return@AsyncFunction "denied"
            runBlocking {
                WearablesManager.checkPermissionStatus(
                    com.meta.wearable.dat.core.types.Permission.CAMERA
                )
            }
        }

        AsyncFunction("requestPermission") { permission: String ->
            if (permission != "camera") throw Exception("Unknown permission: $permission")
            val activity = appContext.currentActivity
                ?: throw Exception("Current activity not available")
            runBlocking {
                WearablesManager.requestPermission(
                    activity,
                    com.meta.wearable.dat.core.types.Permission.CAMERA
                )
            }
        }

        // MARK: - Devices

        AsyncFunction("getDevices") {
            WearablesManager.getDevices()
        }

        AsyncFunction("getDevice") { identifier: String ->
            WearablesManager.getDevice(identifier)
        }

        AsyncFunction("openFirmwareUpdate") {
            val activity = appContext.currentActivity
                ?: throw Exception("Current activity not available")
            WearablesManager.openFirmwareUpdate(activity)
        }

        AsyncFunction("openDATGlassesAppUpdate") {
            val activity = appContext.currentActivity
                ?: throw Exception("Current activity not available")
            WearablesManager.openDATGlassesAppUpdate(activity)
        }

        // MARK: - Session Management

        AsyncFunction("createSession") { deviceId: String? ->
            WearablesManager.createSession(deviceId)
        }

        AsyncFunction("startSession") { sessionId: String ->
            WearablesManager.startSession(sessionId)
        }

        AsyncFunction("stopSession") { sessionId: String ->
            WearablesManager.stopSession(sessionId)
        }

        AsyncFunction("addCameraToSession") { sessionId: String, config: Map<String, Any> ->
            CameraSessionManager.addCameraToSession(sessionId, config)
        }

        AsyncFunction("removeCameraFromSession") { sessionId: String ->
            CameraSessionManager.removeCameraFromSession(sessionId)
        }

        // MARK: - Photo Capture

        // Display

        AsyncFunction("addDisplayToSession") { sessionId: String ->
            DisplayManager.addDisplayToSession(sessionId)
        }

        AsyncFunction("renderDisplay") { sessionId: String, root: Map<String, Any?> ->
            runBlocking {
                DisplayManager.renderDisplay(sessionId, root)
            }
            null
        }

        AsyncFunction("clearDisplay") { sessionId: String ->
            runBlocking {
                DisplayManager.clearDisplay(sessionId)
            }
            null
        }

        AsyncFunction("removeDisplayFromSession") { sessionId: String ->
            DisplayManager.removeDisplayFromSession(sessionId)
        }

        AsyncFunction("getDisplayState") { sessionId: String ->
            DisplayManager.getDisplayState(sessionId) ?: "stopped"
        }

        AsyncFunction("capturePhoto") { format: String ->
            val context = appContext.reactContext?.applicationContext
                ?: throw Exception("Application context not available")
            runBlocking {
                CameraSessionManager.capturePhoto(context, format)
            }
            null
        }

        // MARK: - Mock Device Kit (DEBUG only)

        AsyncFunction("enableMockDeviceKit") { config: Map<String, Any> ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            val context = appContext.reactContext?.applicationContext
                ?: throw Exception("Application context not available")
            val initiallyRegistered = config["initiallyRegistered"] as? Boolean ?: true
            val initialPermissionsGranted = config["initialPermissionsGranted"] as? Boolean ?: true
            MockDeviceManager.enableMockDeviceKit(context, initiallyRegistered, initialPermissionsGranted)
        }

        AsyncFunction("disableMockDeviceKit") {
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            val context = appContext.reactContext?.applicationContext
                ?: throw Exception("Application context not available")
            MockDeviceManager.disableMockDeviceKit(context)
        }

        AsyncFunction("isMockDeviceKitEnabled") {
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            val context = appContext.reactContext?.applicationContext
                ?: throw Exception("Application context not available")
            MockDeviceManager.isMockDeviceKitEnabled(context)
        }

        AsyncFunction("pairMockDevice") { model: String? ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            val context = appContext.reactContext?.applicationContext
                ?: throw Exception("Application context not available")
            MockDeviceManager.pairMockDevice(context, model ?: "rayBanMeta")
        }

        AsyncFunction("unpairMockDevice") { deviceId: String ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            val context = appContext.reactContext?.applicationContext
                ?: throw Exception("Application context not available")
            MockDeviceManager.unpairMockDevice(context, deviceId)
        }

        AsyncFunction("getMockDevices") {
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            MockDeviceManager.getMockDevices()
        }

        AsyncFunction("mockDevicePowerOn") { id: String ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            MockDeviceManager.powerOn(id)
        }

        AsyncFunction("mockDevicePowerOff") { id: String ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            MockDeviceManager.powerOff(id)
        }

        AsyncFunction("mockDeviceDon") { id: String ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            MockDeviceManager.don(id)
        }

        AsyncFunction("mockDeviceDoff") { id: String ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            MockDeviceManager.doff(id)
        }

        AsyncFunction("mockDeviceFold") { id: String ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            MockDeviceManager.fold(id)
        }

        AsyncFunction("mockDeviceUnfold") { id: String ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            MockDeviceManager.unfold(id)
        }

        AsyncFunction("mockDeviceTap") { id: String ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            MockDeviceManager.tap(id)
        }

        AsyncFunction("mockDeviceTapAndHold") { id: String ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            MockDeviceManager.tapAndHold(id)
        }

        AsyncFunction("mockDeviceSetCameraFeed") { id: String, fileUrl: String ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            MockDeviceManager.setCameraFeed(id, fileUrl)
        }

        AsyncFunction("mockDeviceSetCapturedImage") { id: String, fileUrl: String ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            MockDeviceManager.setCapturedImage(id, fileUrl)
        }

        AsyncFunction("mockDeviceSetCameraFeedFromCamera") { id: String, facing: String ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            MockDeviceManager.setCameraFeedFromCamera(id, facing)
        }

        AsyncFunction("mockSetPermissionStatus") { permission: String, status: String ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            val context = appContext.reactContext?.applicationContext
                ?: throw Exception("Application context not available")
            MockDeviceManager.setPermissionStatus(context, permission, status)
        }

        AsyncFunction("mockSetPermissionRequestResult") { permission: String, result: String ->
            if (!isDebug) throw Exception("Mock devices are only available in debug builds")
            val context = appContext.reactContext?.applicationContext
                ?: throw Exception("Application context not available")
            MockDeviceManager.setPermissionRequestResult(context, permission, result)
        }

        // MARK: - View

        View(EMWDATView::class) {
            Prop("isActive") { view: EMWDATView, isActive: Boolean ->
                view.setActive(isActive)
            }

            Prop("resizeMode") { view: EMWDATView, resizeMode: String ->
                view.setResizeMode(resizeMode)
            }
        }
    }
}
